migrate to 2024 edition

jonaspleyer opened the pull request:
This is an initial advancement to increase the edition version to 2024. In particular I have done the following non-structural changes:

- Add the `unsafe` keyword to function calls of unsafe functions
- Mark `extern "C"` blocks as `unsafe extern "C"`

For now, the error messages which are being produced are different to the previous ones. I have commented out the `test_compile_errors` test case. The remaining tests are all passing.
I have not yet added an entry into the `newsfragments` folder but in the process of merging this PR, I can do this as well.

Let me know what you think.

alex commented:
Thanks for working on this.

Since this requires raising our MSRV significantly, this isn't something we'll be able to consider merging in the new future.

It looks like overwhelmingly these changes relate to how `unsafe` scoping works. I believe there are some warnings we can enable on older versions of Rust to get the same behavior? If so, it probably makes sense to see if those can be landed now, independent of raising the edition.

jonaspleyer commented:
There are two lints which the new Rust edition mentions (see https://doc.rust-lang.org/edition-guide/rust-2024/index.html).

```
#![warn(missing_unsafe_on_extern, unsafe_op_in_unsafe_fn)]
```
https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#missing-unsafe-on-extern
https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unsafe-op-in-unsafe-fn

However, it would mean that we would need to apply these warnings for each crate in the workspace individually.

jonaspleyer commented:
> ### Benchmarks breakdown
> 	Benchmark 	`BASE` 	`HEAD` 	Change
> ❌ 	`extract_hashbrown_set` 	16.1 ms 	18.4 ms 	-12.62%

Since my changes have only been semantic, and not changed any functionality or versions of dependecies, this should not have introduced a regression this big. Either the test was an outlier or something else is happening when switching to the new edition and a new compiler version.

jonaspleyer commented:
> Hmm, is there some way to use workspace settings to enable those warnings universally?

I am not sure about this. I have opened a thread on the users.rust-lang.org forum to see if anyone has a idea on how to do this.
https://users.rust-lang.org/t/warnings-on-workspace-level/126967
I will keep you updated here.

davidhewitt commented:
We should be able to configure the lints via the `[workspace.lints.rust]` setting already present in our `Cargo.toml`.

jonaspleyer commented:
I have rebased the branch to only include the following commits addressing the previously mentioned comments:

- a4c52df0622745ace6965c64b7308a652a8af54b mark ffi extern blocks and functions as unsafe
- 866a6c50f50573b6a43ee583c38e562c61c95e5e mark unsafe function calls with keyword
- e574347c9967749f323650a5740f01cc9c9d82c5 mark unsafe function calls in tests as well

> Hmm, is there some way to use workspace settings to enable those warnings
universally?

> We should be able to configure the lints via the [workspace.lints.rust] setting already present in our Cargo.toml.

- adf36854f3d0ad54f51c925f86eabd15a7ef2ab1 add workspace lints to enhance migration to 2024 edition

davidhewitt commented:
Thanks for moving forward with this. `cargo fmt` needs to be run else the CI won't try further.

I had a quick test and on our MSRV (1.63) `unsafe extern` is not valid syntax:

```
error: extern block cannot be declared unsafe
  --> pyo3-ffi/src/abstract_.rs:25:1
   |
25 | unsafe extern "C" {
   | ^^^^^^
```

... so we may have to drop https://github.com/PyO3/pyo3/commit/a4c52df0622745ace6965c64b7308a652a8af54b from this PR.

jonaspleyer commented:
> ... so we may have to drop [a4c52df](https://github.com/PyO3/pyo3/commit/a4c52df0622745ace6965c64b7308a652a8af54b) from this PR.

I have updated the PR and added another small commit for formatting

davidhewitt commented on pyo3-ffi/src/abstract_.rs:
```diff
@@ -83,7 +83,7 @@ extern "C" {
     1 << (8 * std::mem::size_of::<size_t>() as size_t - 1);
 
-extern "C" {
+unsafe extern "C" {
```
```suggestion
extern "C" {
```

davidhewitt commented on pyo3-ffi/src/pystate.rs:
```diff
@@ -67,7 +67,7 @@ extern "C" {
 // skipped non-limited / 3.9 PyThreadState_GetID
 
-extern "C" {
+unsafe extern "C" {
```
```suggestion
extern "C" {
```

davidhewitt reviewed (approved):
Thanks, this is essentially now purely changing the code to have `unsafe { }` blocks inside `unsafe fn` and enforcing that as per 2024 edition best practice.

Let's get this merged; I think it's gonna be super easy for this to conflict with everything so better get in 😬

jonaspleyer commented:
It seems that the python3.12-x64 build with rust `1.63` fails. I am not sure how to proceed.

davidhewitt commented:
Ungh sorry I fat-finger pushed to this branch and now can't reopen, continued in #4994
